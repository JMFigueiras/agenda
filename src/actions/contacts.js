import constant from 'lodash/constant';

export const FETCH_CONTACTS_REQUESTED = 'FETCH_CONTACTS_REQUESTED';
export const FETCH_CONTACTS_SUCCEEDED = 'FETCH_CONTACTS_SUCCEEDED';
export const SUBMIT_CONTACT_DATA_REQUESTED = 'SUBMIT_CONTACT_DATA_REQUESTED';
export const SUBMIT_CONTACT_DATA_SUCCEEDED = 'SUBMIT_CONTACT_DATA_SUCCEEDED';

export const UPDATE_CONTACT_DATA = 'UPDATE_CONTACT_DATA';

export const FETCH_CONTACT_REQUESTED = 'FETCH_CONTACT_REQUESTED';

export const fetchContactsRequested = constant({type: FETCH_CONTACTS_REQUESTED});
export const fetchContactsSucceeded = contacts => ({type: FETCH_CONTACTS_SUCCEEDED, contacts});
export const submitContactDataRequested = constant({type: SUBMIT_CONTACT_DATA_REQUESTED});
export const submitContactDataSucceded = constant({type: SUBMIT_CONTACT_DATA_SUCCEEDED});
export const updateContactData = contact => ({type: UPDATE_CONTACT_DATA, contact});

export const fetchContactRequested = id => ({type: FETCH_CONTACT_REQUESTED, id});
