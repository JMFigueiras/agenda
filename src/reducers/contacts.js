import {
    FETCH_CONTACTS_SUCCEEDED,
    SUBMIT_CONTACT_DATA_SUCCEEDED,
    UPDATE_CONTACT_DATA
} from '@actions/contacts';

export default (state = {contacts: [], contact: {}}, action) => {
    switch (action.type) {
        case FETCH_CONTACTS_SUCCEEDED:
            return {...state, contacts: action.contacts};
        case SUBMIT_CONTACT_DATA_SUCCEEDED:
            return {...state, success: true, contact: {}};
        case UPDATE_CONTACT_DATA:
            return {...state, contact: action.contact};
        default:
            return state;
    }
};
