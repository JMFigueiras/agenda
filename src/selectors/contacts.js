/* eslint-disable import/prefer-default-export */
export const getContacts = name => ({contacts: {contacts}}) => contacts;
