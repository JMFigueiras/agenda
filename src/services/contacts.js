import Http from './http';

const API = 'api/contacts';

class Contacts {
    static fetchContacts() {
        return Http.get(API);
    }

    static fetchContact(id) {
        return Http.get(`${API}/${id}`);
    }

    static submitContact(contact) {
        if (!contact.id) {
            return Http.post(API, {...contact});
        }
        return Http.put(`${API}/${contact.id}`, {...contact});
    }
}

export default Contacts;
