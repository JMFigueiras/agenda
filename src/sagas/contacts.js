import {call, put, select} from 'redux-saga/effects';
import {fetchContactsSucceeded, submitContactDataSucceded, updateContactData} from '@actions/contacts';
import ContactsService from '@services/contacts';


// eslint-disable-next-line import/prefer-default-export
export function* fetchContacts() {
    const contacts = yield call(ContactsService.fetchContacts);
    yield put(
        fetchContactsSucceeded(contacts)
    );
}

export function* fetchContact({id}) {
    const contact = yield call(ContactsService.fetchContact, id);
    yield put(
        updateContactData(contact)
    );
}
export function* submitContactData() {
    const contact = yield select(state => state.contacts.contact);
    const result = yield call(ContactsService.submitContact, contact);
    if (result.success) {
        yield put(
            submitContactDataSucceded()
        );
    }
}
