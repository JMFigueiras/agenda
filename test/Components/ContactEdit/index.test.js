import React from 'react';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import {MemoryRouter} from 'react-router';
import Enzyme, {mount} from 'enzyme';
import {Input} from 'reactstrap';
import forEach from 'lodash/forEach';

import ContactEdit from '@components/Contact/Edit';
import {UPDATE_CONTACT_DATA} from '@actions/contacts';

import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({adapter: new Adapter()});

const mockStore = configureStore([]);

describe('ContactEdit Component Snapshot', () => {
    let store;
    let component;
    let contact;
    const match = {
        params: {
            id: null
        }
    };

    beforeEach(() => {
        contact = {
            firstName: '',
            lastName: '',
            email: '',
            gender: '',
            birthDate: '',
            phoneNumber: '',
            address: '',
            role: '',
            notes: ''
        };

        store = mockStore({
            contacts: {
                contact
            }
        });
    });

    test('should mount Component', () => {
        const snapshot = renderer.create(
            <Provider store={store}>
                <MemoryRouter>
                    <ContactEdit {...{match}}/>
                </MemoryRouter>
            </Provider>
        ).toJSON();

        expect(snapshot).toMatchSnapshot();
    });

    function mountComponent() {
        return mount(
            <Provider store={store}>
                <MemoryRouter>
                    <ContactEdit {...{match}}/>
                </MemoryRouter>
            </Provider>
        );
    }

    it('should render with given state from store', () => {
        component = mountComponent();
        forEach([
            'firstName',
            'lastName',
            'email',
            'gender',
            'birthDate',
            'phoneNumber',
            'address',
            'role',
            'notes'
        ], field => {
            const inputField = component.find(Input).find({name: field});
            expect(inputField.exists()).toBe(true);
        });
    });

    it('should call onChange event for firstName with the expected params', () => {
        component = mountComponent();
        const inputField = component.find(Input).find({name: 'firstName'});
        expect(inputField.exists()).toBe(true);
        // Busco el Ãºltimo componente, (el hijo)
        inputField.last().simulate('change', {target: {value: 'fake-firstName'}});
        expect(store.getActions()).toEqual([
            {
                type: UPDATE_CONTACT_DATA,
                contact: {
                    ...contact,
                    firstName: 'fake-firstName'
                }
            }
        ]);
    });

    it('should call onChange event for lastName with the expected params', () => {
        component = mountComponent();
        const inputField = component.find(Input).find({name: 'lastName'});
        expect(inputField.exists()).toBe(true);
        inputField.last().simulate('change', {target: {value: 'fake'}});
        expect(store.getActions()).toEqual([
            {
                type: UPDATE_CONTACT_DATA,
                contact: {
                    ...contact,
                    lastName: 'fake'
                }
            }
        ]);

        inputField.last().simulate('change', {target: {value: 'fake-lastName'}});
        expect(store.getActions()).toEqual([
            {
                type: UPDATE_CONTACT_DATA,
                contact: {
                    ...contact,
                    lastName: 'fake'
                }
            },
            {
                type: UPDATE_CONTACT_DATA,
                contact: {
                    ...contact,
                    lastName: 'fake-lastName'
                }
            }
        ]);
    });
});
