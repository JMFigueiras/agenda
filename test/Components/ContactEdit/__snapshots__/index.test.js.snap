// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactEdit Component Snapshot should mount Component 1`] = `
<div
  className="container-fluid"
>
  <form
    className=""
  >
    <div
      className="form-group"
    >
      <label
        className=""
      >
        Nombre
        <br />
        <input
          className="form-control"
          control="firstName"
          label="Nombre"
          name="firstName"
          onChange={[Function]}
          path="firstName"
          type="text"
          value=""
        />
      </label>
    </div>
    <div
      className="form-group"
    >
      <label
        className=""
      >
        Apellido
        <br />
        <input
          className="form-control"
          control="lastName"
          label="Apellido"
          name="lastName"
          onChange={[Function]}
          path="lastName"
          type="text"
          value=""
        />
      </label>
    </div>
    <div
      className="form-group"
    >
      <label
        className=""
      >
        Email
        <br />
        <input
          className="form-control"
          control="email"
          label="Email"
          name="email"
          onChange={[Function]}
          path="email"
          type="text"
          value=""
        />
      </label>
    </div>
    <div
      className="form-group"
    >
      <label
        className=""
      >
        Sexo
        <br />
        <input
          className="form-control"
          control="gender"
          label="Sexo"
          name="gender"
          onChange={[Function]}
          path="gender"
          type="text"
          value=""
        />
      </label>
    </div>
    <div
      className="form-group"
    >
      <label
        className=""
      >
        Fecha de nacimiento
        <br />
        <input
          className="form-control"
          control="birthDate"
          label="Fecha de nacimiento"
          name="birthDate"
          onChange={[Function]}
          path="birthDate"
          type="date"
          value=""
        />
      </label>
    </div>
    <div
      className="form-group"
    >
      <label
        className=""
      >
        Número de teléfono
        <br />
        <input
          className="form-control"
          control="phoneNumber"
          label="Número de teléfono"
          name="phoneNumber"
          onChange={[Function]}
          path="phoneNumber"
          type="phone"
          value=""
        />
      </label>
    </div>
    <div
      className="form-group"
    >
      <label
        className=""
      >
        Dirección
        <br />
        <input
          className="form-control"
          control="address"
          label="Dirección"
          name="address"
          onChange={[Function]}
          path="address"
          type="phone"
          value=""
        />
      </label>
    </div>
    <div
      className="form-group"
    >
      <label
        className=""
      >
        Cargo
        <br />
        <input
          className="form-control"
          control="role"
          label="Cargo"
          name="role"
          onChange={[Function]}
          path="role"
          type="text"
          value=""
        />
      </label>
    </div>
    <div
      className="form-group"
    >
      <label
        className=""
      >
        Notas
        <br />
        <input
          className="form-control"
          control="notes"
          label="Notas"
          name="notes"
          onChange={[Function]}
          path="notes"
          type="text"
          value=""
        />
      </label>
    </div>
    <button
      aria-label={null}
      className="btn btn-secondary"
      onClick={[Function]}
      type="button"
    >
      Submit
    </button>
  </form>
</div>
`;
